### PR TITLE
The problem in the pull request $67 has been solved, hopefully.

### PR DESCRIPTION
**Pull request by onurcanbektas**
To make sure, I have tryied removing various items from character, and didn't saw any problem.

### CODE REVIEW

**Comment by twostars** [comment]
So... okay, I think I need to explain why the packet is structured as it is. And note: the client source still doesn't actually implement all of it.

In the server we have this:
```
{
	Packet result(WIZ_ITEM_COUNT_CHANGE);

	result << uint16_t(1);
	result << uint8_t(1);
	result << uint8_t(bPos);
	result << nItemID << nCount;
	result << uint8_t(bNewItem ? 100 : 0);
	result << sDurability;
	if (Time != 0)
	{
		result << Time << Time;
	}

	SetUserAbility(false);
	SendItemWeight();

	Send(&result);
}
```
Ignoring the irrelevant time stuff (which is not accurate, anyway, but isn't part of the 1.298 structure regardless), take special note of the extra pieces of information we're adding in here:
```
	result << uint8_t(bNewItem ? 100 : 0);
	result << sDurability;
```

The flag to indicate a new item is used by the client exactly for this reason: it's used for showing the notification, directly in the handler. The client doesn't need extra logic to determine if it's feasible -- this is entirely up to the server to determine. The client should merely do as it's told.

How do I know they do this? I've disassembled the client to confirm this. You can see MsgRecv_ItemCountChange() here:
```
int __thiscall sub_5D0740(_DWORD *this, int a2, int *a3)
{
  int v3; // esi@1
  int *v4; // edi@1
  _DWORD *v5; // ebx@1
  int v6; // edx@1
  int result; // eax@2
  int v8; // edx@7
  int v9; // ecx@13
  signed int v10; // ebx@14
  int v11; // ecx@22
  __int16 v12; // ax@23
  int v13; // ebp@25
  int v14; // edi@25
  signed int v15; // eax@26
  int v16; // esi@29
  int v17; // edi@29
  int v18; // ebp@30
  unsigned __int8 v19; // cl@31
  int v20; // esi@32
  char v21; // al@36
  char *v22; // edi@38
  char *v23; // eax@40
  char *v24; // eax@43
  char *v25; // ecx@45
  unsigned int v26; // edi@48
  int v27; // esi@51
  int v28; // esi@60
  int v29; // esi@61
  char *v30; // eax@62
  char *v31; // ecx@64
  unsigned int v32; // esi@66
  int v33; // edi@69
  char v34; // [sp+Fh] [bp-269h]@0
  _DWORD *v35; // [sp+10h] [bp-268h]@1
  int v36; // [sp+14h] [bp-264h]@22
  int v37; // [sp+18h] [bp-260h]@10
  char v38; // [sp+1Ch] [bp-25Ch]@35
  char *v39; // [sp+20h] [bp-258h]@40
  unsigned int v40; // [sp+24h] [bp-254h]@48
  char v41; // [sp+2Ch] [bp-24Ch]@62
  char *v42; // [sp+30h] [bp-248h]@64
  unsigned int v43; // [sp+34h] [bp-244h]@66
  int v44; // [sp+3Ch] [bp-23Ch]@17
  int v45; // [sp+40h] [bp-238h]@20
  int v46; // [sp+44h] [bp-234h]@5
  int v47; // [sp+48h] [bp-230h]@8
  int v48; // [sp+4Ch] [bp-22Ch]@11
  int v49; // [sp+50h] [bp-228h]@13
  int v50; // [sp+54h] [bp-224h]@25
  int v51; // [sp+58h] [bp-220h]@29
  int v52; // [sp+5Ch] [bp-21Ch]@32
  int v53; // [sp+60h] [bp-218h]@32
  int v54; // [sp+64h] [bp-214h]@60
  int v55; // [sp+68h] [bp-210h]@25
  char v56; // [sp+6Ch] [bp-20Ch]@42
  char v57; // [sp+16Ch] [bp-10Ch]@66
  int v58; // [sp+274h] [bp-4h]@35

  v3 = a2;
  v4 = a3;
  v5 = this;
  v35 = this;
  v6 = *a3 + 2;
  *a3 = v6;
  if ( v6 <= *(_DWORD *)(a2 + 4) )
    LOWORD(result) = *(_WORD *)(*(_DWORD *)(a2 + 8) + v6 - 2);
  else
    LOWORD(result) = 0;
  result = (signed __int16)result;
  if ( (signed __int16)result > 0 )
  {
    v46 = result;
    while ( 1 )
    {
      v8 = *v4 + 1;
      *v4 = v8;
      LOBYTE(v47) = v8 <= *(_DWORD *)(v3 + 4) ? *(_BYTE *)(*(_DWORD *)(v3 + 8) + v8 - 1) : 0;
      v37 = (unsigned __int8)v47;
      *v4 = v8 + 1;
      LOBYTE(v48) = v8 + 1 <= *(_DWORD *)(v3 + 4) ? *(_BYTE *)(*(_DWORD *)(v3 + 8) + v8 + 1 - 1) : 0;
      v9 = (unsigned __int8)v48;
      *v4 = v8 + 5;
      v49 = v9;
      v10 = v8 + 5 <= *(_DWORD *)(v3 + 4) ? *(_DWORD *)(*(_DWORD *)(v3 + 8) + v8 + 5 - 4) : 0;
      *v4 = v8 + 9;
      v44 = v8 + 9 <= *(_DWORD *)(v3 + 4) ? *(_DWORD *)(*(_DWORD *)(v3 + 8) + v8 + 9 - 4) : 0;
      *v4 = v8 + 10;
      LOBYTE(v45) = v8 + 10 <= *(_DWORD *)(v3 + 4) ? *(_BYTE *)(*(_DWORD *)(v3 + 8) + v8 + 10 - 1) : 0;
      v11 = (unsigned __int8)v45;
      *v4 = v8 + 12;
      v36 = v11;
      v12 = v8 + 12 <= *(_DWORD *)(v3 + 4) ? *(_WORD *)(*(_DWORD *)(v3 + 8) + v8 + 12 - 2) : 0;
      v13 = v12;
      v14 = dword_8186F4;
      v55 = 1000 * (v10 / 1000);
      sub_4532A0(&v50, &v55);
      v15 = v50 == *(_DWORD *)(v14 + 24) ? 0 : v50 + 16;
      if ( v37 == 2 )
      {
        v37 = 1000 * (v10 / 1000);
        v16 = dword_8186F4;
        v17 = 0;
        sub_4532A0(&v51, &v37);
        if ( v51 != *(_DWORD *)(v16 + 24) )
        {
          v18 = v51 + 16;
          if ( v51 != -16 )
          {
            v19 = *(_BYTE *)(v51 + 20);
            if ( v19 < 0x18u )
            {
              v20 = dword_8186F8[*(_BYTE *)(v51 + 20)];
              v53 = v10 % 1000;
              sub_453310(&v52, &v53);
              if ( v52 == *(_DWORD *)(v20 + 24) )
                v17 = 0;
              else
                v17 = v52 + 16;
            }
            v38 = v34;
            std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Tidy(0);
            v58 = 0;
            _LoadStringFromResource(6092, (int)&v38);
            if ( v17 && ((v21 = *(_BYTE *)(v17 + 52), v21 == 4) || v21 == 6) )
            {
              v22 = *(char **)(v17 + 8);
              if ( !v22 )
                v22 = byte_6A3560;
              v23 = v39;
              if ( !v39 )
                v23 = byte_6A3560;
              sprintf(&v56, v23, v22);
            }
            else
            {
              v24 = *(char **)(v18 + 12);
              if ( !v24 )
                v24 = byte_6A3560;
              v25 = v39;
              if ( !v39 )
                v25 = byte_6A3560;
              sprintf(&v56, v25, v24);
            }
            v26 = dword_6A3564;
            std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Split(&v38);
            if ( v40 < v26 )
              v26 = v40;
            if ( v26 )
            {
              memcpy(v39, &v39[v26], v40 - v26);
              v27 = v40 - v26;
              if ( sub_403350((int)&v38, v40 - v26, 0) )
                std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Eos(v27);
            }
            sub_4083A0(&v56, strlen(&v56));
            if ( v35[18] )
              sub_52ACD0(&v38, -256);
            v58 = -1;
            std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Tidy(1);
          }
        }
      }
      else
      {
        if ( *(_BYTE *)(v15 + 64) == -1 )
          sub_4A2A60(v35[15], v37, v49, v13, v10);
        else
          sub_4A2A60(v35[15], v37, v49, v44, v10);
        if ( v36 == 100 )
        {
          v36 = 1000 * (v10 / 1000);
          v28 = dword_8186F4;
          sub_4532A0(&v54, &v36);
          if ( v54 != *(_DWORD *)(v28 + 24) )
          {
            v29 = v54 + 16;
            if ( v54 != -16 )
            {
              v41 = v34;
              std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Tidy(0);
              v58 = 1;
              _LoadStringFromResource(7613, (int)&v41);
              v30 = *(char **)(v29 + 12);
              if ( !v30 )
                v30 = byte_6A3560;
              v31 = v42;
              if ( !v42 )
                v31 = byte_6A3560;
              sprintf(&v57, v31, v30);
              v32 = dword_6A3564;
              std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Split(&v41);
              if ( v43 < v32 )
                v32 = v43;
              if ( v32 )
              {
                memcpy(v42, &v42[v32], v43 - v32);
                v33 = v43 - v32;
                if ( sub_403350((int)&v41, v43 - v32, 0) )
                  std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Eos(v33);
              }
              sub_4083A0(&v57, strlen(&v57));
              if ( v35[18] )
                sub_52ACD0(&v41, -256);
              v58 = -1;
              std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Tidy(1);
            }
          }
        }
      }
      result = v46-- - 1;
      if ( !v46 )
        break;
      v4 = a3;
      v3 = a2;
    }
    v5 = v35;
  }
  if ( v5[27] )
    result = sub_4E2D10();
  if ( v5[28] )
    result = sub_4DF360();
  return result;
}
```
Note the checks on 100; that's the flag being sent (that we're not even reading in the client).

All of this extra logic and the working around of it is confusing, goes against the intention of the method itself and is honestly fairly untrustworthy.

What I'd do is simplify all of this immensely by:
1. Reverting the changes to ItemCountChange() (in the original commit as well)
and 2. Reading in the extra data from the packet and handling the notification based on this, instead.

This removes any and all risk of breaking things in ItemCountChange() and ensures consistency/accuracy, whilst simplifying this logic immensely. :)

**Comment by SavvyNik** [comment]
Great explanation. What program are you using to disassemble clients?

**Comment by twostars** [comment]
The above is via IDA's hex rays plugin.

**Comment by onurcanbektas** [comment]
@twostars  Thanks for this detailed comment. However, I have a question. For example, if in my inventory I have , let say, 1 potion of wisdom, then if I receive one more PoW, will this flag again be set as true ?

**Comment by twostars** [comment]
"New items" are considered items in new slots, because you merely update the stack size of existing items. 
You can find this out yourself from the server, however.

**Comment by twostars** [comment]
Let me try again, since my response was overly confusing.
Basically: yes. The flag exists only for that notification. It's set only when giving items (not taking them).

**Comment by onurcanbektas** [comment]
@twostars  Got it, thanks.
